force-app/main/default/classes/COFF_Affiliation_TDTM.cls

```diff
@@ -325,119 +364,189 @@ public class COFF_Affiliation_TDTM extends TDTM_Runnable {
-                                                                        Affiliation__c = oldConns[i].Affiliation__c,
-                                                                        Primary__c = true);
-                if (UTIL_Describe.isValidField(UTIL_Namespace.StrTokenNSPrefix('Course_Enrollment__c'),'RecordTypeId')) {
-                    conn.put('RecordTypeId', facultyRecordTypeId);
+        subQueryString += ')';
+        String queryString = 'SELECT Id, Faculty__c';
+        queryString += subQueryString;
+        queryString += ' FROM Course_Offering__c WHERE Id IN :offsWithChangedFaculty';
+
+        List<Course_Offering__c> updatedCourseOfferings = Database.query(queryString);
+
+        for (Course_Offering__c updatedCourseOffering : updatedCourseOfferings) {
+            Boolean hasExistingCourseConnection = false;
+
+            if (updatedCourseOffering.Faculty__c == null) {
+                continue;
+            }
+
+            if (updatedCourseOffering.Course_Enrollment__r.size() == 0) {
+                continue;
+            }
+
+            for (Course_Enrollment__c childCourseConnection : updatedCourseOffering.Course_Enrollment__r) {
+                if (updatedCourseOffering.Faculty__c == childCourseConnection.Contact__c) {
+                    if (childCourseConnection.Primary__c == false) {
+                        childCourseConnection.Primary__c = true;
+                        objectsToUpdate.add((SObject) childCourseConnection);
+                    }
+                    hasExistingCourseConnection = true;
+                } else {
+                    if (childCourseConnection.Primary__c == true) {
+                        childCourseConnection.Primary__c = false;
+                        objectsToUpdate.add((SObject) childCourseConnection);
+
+                        if (
+                            updatedCouserEnrollmentBycourseOfferingId.keySet().contains(updatedCourseOffering.Id) ==
+                            false
+                        ) {
+                            updatedCouserEnrollmentBycourseOfferingId.put(
+                                updatedCourseOffering.Id,
+                                childCourseConnection
+                            );
+                        }
+                    }
                 }
-                objectsToInsert.add((SObject)conn);
             }
 
-            Course_Enrollment__c oldConn = oldConns[i];
-            if (oldConn.Primary__c) {
-                oldConn.Primary__c = false;
-                objectsToUpdate.add((SObject)oldConn);
+            if (hasExistingCourseConnection == false) {
+                Course_Enrollment__c newCourseConnection = new Course_Enrollment__c(
+                    Contact__c = updatedCourseOffering.Faculty__c,
+                    Primary__c = true
+                );
+
+                if (
```

Comment:
Starting with this line is where we can get into trouble by assuming there was only one former primary course connection.